cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -1107,17 +1107,9 @@ public void removeApplication(ApplicationReference appRef)
     // since already checked DELETE privileges above, no need to check again for appDetail
     // ensure no running programs across all versions of the application
     ensureNoRunningPrograms(appRef);
```

Comment:
this is an existing issue, but we should actually be checking this in the same transaction that we delete the app, otherwise there is a race condition. Can you open a jira for this?

Reply:
this implementation is also not as performant as it should be, it loads all active runs into an in-memory map then checks if the map is empty. When moving it inside the AppMetadataStore, we should be doing a scan with limit 1 instead of getting everything.

Reply:
that can be addressed in this PR? Created https://cdap.atlassian.net/browse/CDAP-20423

Reply:
should do it in a separate PR since it's unrelated and an existing problem.